Ready-to-Run support for generic methods without generic dictionary

JohnChen0 opened the pull request:
Add support in Ready-to-Run images for generic methods that do not use generic dictionary. The change is feature complete, but still needs more testing.

JohnChen0 commented:
@jkotas @fadimounir PTAL

jkotas commented on src/inc/corcompile.h:
```diff
@@ -1863,6 +1863,8 @@ class ICorCompileInfo
 
     virtual BOOL AreAllClassesFullyLoaded(CORINFO_MODULE_HANDLE moduleHandle) = 0;
+
+    virtual int GetStableMethodHashCode(CORINFO_METHOD_HANDLE hMethod) = 0;
```
Call this `GetVersionResilientMethodHashCode` to make it distinct from the existing stable hashcode?

jkotas commented on src/vm/stablehashcode.cpp:
```diff
@@ -0,0 +1,79 @@
+#include "typehashingalgorithms.h"
+
+int GetStableHashCode(TypeHandle type)
+{
```
Could you please reconcile it with the hashcode computation added in earlier Fadi's checkin (check ZapImage::OutputTypesTableForReadyToRun in https://github.com/dotnet/coreclr/pull/3735)?

jkotas commented on src/vm/typehashingalgorithms.h:
```diff
@@ -0,0 +1,85 @@
+
+    for (COUNT_T i = 0; i < src.GetCount(); i += 2)
+    {
+        hash1 = (hash1 + _rotl(hash1, 5)) ^ src[i];
```
We should base the hashcode on UTF8 strings to avoid the cost of coverting to Unicode - like what is in Fadi's checkin.

jkotas replied:
It may be a good idea to mark all places where we do something different from current CoreRT with a "// DIFFERENT FROM CORERT" comment.

jkotas commented on src/vm/typehashingalgorithms.h:
```diff
@@ -0,0 +1,85 @@
+        {
+        case 2: typeName = W("System.MDArrayRank2`1");
+        case 3: typeName = W("System.MDArrayRank3`1");
+        default: _ASSERTE(!"NYI - MD arrays with rank > 3"); typeName = NULL;
```
At minimum, we need to support all ranks supported by CoreCLR (I think it is up to 15). (Or maybe do something from CoreRT.)

mikedn replied:
> At minimum, we need to support all ranks supported by CoreCLR (I think it is up to 15). 

AFAIK it's 32, I looked into this when doing the initialize array PR.

jkotas commented on src/vm/readytoruninfo.cpp:
```diff
@@ -469,22 +477,102 @@ ReadyToRunInfo::ReadyToRunInfo(Module * pModule, PEImageLayout * pLayout, READYT
-        return NULL;
+    BYTE kind = *pSig++;
+    _ASSERTE(kind == ENCODE_METHOD_ENTRY);
+    if (kind != ENCODE_METHOD_ENTRY)
```
I do not think we need `ENCODE_METHOD_ENTRY` prefix in the signatures.

jkotas replied:
(call `GetCompileInfo()->EncodeMethod` directly to encode it)

JohnChen0 commented:
Fixed some bugs found during testing, and updated to address review comments. Have successfully generated Ready-to-Run images for CLI assemblies, and verified that CLI still works correctly.

jkotas commented on src/vm/compile.cpp:
```diff
@@ -76,6 +76,8 @@
 
 #include "argdestination.h"
 
+#include "stablehashcode.h"
```
stablehashcode -> versionresilienthashcode ?

jkotas commented on src/vm/stablehashcode.cpp:
```diff
@@ -0,0 +1,137 @@
+#include "stablehashcode.h"
+#include "typehashingalgorithms.h"
+
+// DIFFERENT FROM CORERT: Hashing code currently different from CoreRT
```
Could you please add this comment to the specific lines that differ? I believe it should be mostly the same ... there are just a few details that differ.

jkotas commented on src/vm/typehashingalgorithms.h:
```diff
@@ -0,0 +1,90 @@
+#pragma once
+#include <stdlib.h>
+
+// DIFFERENT FROM CORERT: Hashing code currently different from CoreRT
```
Same here.

jkotas commented on src/vm/stablehashcode.cpp:
```diff
@@ -0,0 +1,137 @@
+        {
+            return ComputeArrayTypeHashCode(GetVersionResilientTypeHashCode(pMT->GetApproxArrayElementTypeHandle()), pMT->GetRank());
+        }
+        else if (!IsNilToken(pMT->GetCl()))
```
We should always get a nice type with metadata here. Do we really need the IsArray check (arrays should come as typedescs here) and the null token check?

jkotas commented on src/vm/stablehashcode.cpp:
```diff
@@ -0,0 +1,137 @@
+        }
+    }
+    else
+    if (type.IsGenericVariable())
```
Do we need this for anything? It should not be needed.

JohnChen0 replied:
It's not needed. Just some bad copy and paste. I'll remove it.

jkotas commented on src/vm/typehashingalgorithms.h:
```diff
@@ -0,0 +1,90 @@
+
+inline static DWORD ComputeArrayTypeHashCode(DWORD elementTypeHashcode, DWORD rank)
+{
+    // Arrays are treated as generic types in some parts of our system. The array hashcodes are 
```
This comment is not true because of the algorithm is different from CoreRT. Delete it and just add comment that say that it is different from CoreRT.

I would change the whole implementation to be just: `return 0xd5313556 + rank;` (ie return same hashcode for single dimensional arrays as CoreRT, but different one for multidimensional arrays).

jkotas commented on src/vm/typehashingalgorithms.h:
```diff
@@ -9,19 +9,26 @@
 // Returns the hashcode value of the 'src' string
 //
-inline static int ComputeNameHashCode(const SString &src)
+inline static DWORD ComputeNameHashCode(LPCUTF8 src)
```
Why did you have to change this to DWORD? I think it is better to keep it as int.

jkotas commented on src/vm/typehashingalgorithms.h:
```diff
@@ -0,0 +1,90 @@
+//
+// Returns the hashcode value of the 'src' string
+//
+inline static DWORD ComputeNameHashCode(LPCUTF8 src)
```
Why did you have to change this to DWORD? I think it is better to keep it as int. It better to avoid Windows specific types in files that potential to be reused between projects.

JohnChen0 replied:
Existing code was using a variety of types to store hash code: `ReadyToRunInfo::TryLookupTypeTokenFromName` used DWORD, some other code used unsigned int, while some used int. I tried to standardize on a single type in all the code involved in this feature, but looks like DWORD was a bad choice. I'll switch to int.

jkotas commented on src/vm/zapsig.cpp:
```diff
@@ -441,6 +441,12 @@ BOOL ZapSig::GetSignatureForTypeHandle(TypeHandle      handle,
+            if (TypeFromToken(tk) == mdtTypeRef)
+            {
+                ENABLE_FORBID_GC_LOADER_USE_IN_THIS_SCOPE();
+                TypeHandle actualHandle = ClassLoader::LoadTypeDefOrRefThrowing(pModule, tk, ClassLoader::ReturnNullIfNotFound, ClassLoader::FailIfUninstDefOrRef, tdAllTypes);
```
This should use ResolveTokenToTypeDefThrowing to avoid doing full load of a wrong type. And we may do try/catch around it and swallow any exceptions - in case there is a problem resolving token for a wrong type.

jkotas replied:
This should use `ResolveTokenToTypeDefThrowing` to avoid full load of a wrong type. Maybe we should also have try/catch around ResolveTokenToTypeDefThrowing here to swallow exceptions caused by resolving token for a wrong type.

jkotas commented on src/vm/zapsig.cpp:
```diff
@@ -487,6 +493,13 @@ BOOL ZapSig::GetSignatureForTypeHandle(TypeHandle      handle,
+            if (TypeFromToken(tk) == mdtTypeRef)
+            {
+                ENABLE_FORBID_GC_LOADER_USE_IN_THIS_SCOPE();
+                TypeHandle actualHandle = ClassLoader::LoadTypeDefOrRefThrowing(pModule, tk, ClassLoader::ReturnNullIfNotFound, ClassLoader::PermitUninstDefOrRef, tdAllTypes);
```
Same here.

jkotas commented on src/zap/zapreadytorun.cpp:
```diff
@@ -175,16 +175,38 @@ class EntryPointVertex : public NativeFormat::Vertex
+
+    virtual void Save(NativeWriter * pWriter)
+    {
+        pWriter->WriteUnsigned(m_pBlob->GetSize());
```
It should not be necessary to encode the size - we do not do it anywhere else for the native signatures.

jkotas commented on src/vm/stablehashcode.cpp:
```diff
@@ -0,0 +1,137 @@
+            break;
+
+        default:
+            return hashcode;
```
Throw for the default case?

jkotas commented on src/vm/stablehashcode.cpp:
```diff
@@ -0,0 +1,137 @@
+
+    while (true)
+    {
+        if (IsNilToken(token))
```
Throw for the null token instead?

jkotas commented on src/vm/stablehashcode.cpp:
```diff
@@ -0,0 +1,137 @@
+        case mdtTypeDef:
+            if (FAILED(pMDImport->GetNameOfTypeDef(token, &szName, &szNamespace)))
+                ThrowHR(COR_E_BADIMAGEFORMAT);
+            if (FAILED(pMDImport->GetNestedClassProps(token, &token)))
```
Should we check for specific HRESULT and throw for everything else?

jkotas commented on src/inc/readytorun.h:
```diff
@@ -55,6 +55,7 @@ enum ReadyToRunSectionType
     READYTORUN_SECTION_DEBUG_INFO                   = 105,
     READYTORUN_SECTION_DELAYLOAD_METHODCALL_THUNKS  = 106,
     READYTORUN_SECTION_AVAILABLE_TYPES              = 107,
```
The earlier `READYTORUN_SECTION_AVAILABLE_TYPES` is not going to be compatible with the current one because of the hash algorithm changed. Could you please change `107` to be unassigned, and make:
```
READYTORUN_SECTION_AVAILABLE_TYPES              = 108,
READYTORUN_SECTION_INSTANCE_METHOD_ENTRYPOINTS  = 109,
```

jkotas commented:
LGTM otherwise

JohnChen0 commented:
@dotnet-bot OSX x64 Checked Build and Test please

JohnChen0 commented:
@dotnet-bot test OSX x64 Checked Build and Test please
